File: ACL_TensorFlow/contrib/cv/FasterRCNN_for_ACL/Benchmark/post_process.cpp
/* *
* Copyright 2020 Huawei Technologies Co., Ltd
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* */

#include "post_process.h"
#include "util.h"
#include <vector>
#include <cstring>
#include <memory>
#include <fstream>
#include "stdio.h"
#include <sys/time.h>
#include <unistd.h>
#include <time.h> 
#include <dirent.h> 
#include <stdarg.h>
#include <iostream>

#include <unistd.h>
#include <thread>
#include <algorithm>
#include <libgen.h>
#include <string.h>
#include <getopt.h>
#include <map>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sstream>
#include <cerrno> 
#include <errno.h> 
#include <sys/errno.h>
#include <unordered_map>


extern int processedCnt;

extern Config cfg;
extern DataFrame outputDataframe;
extern aclError ret;
int topNum = 5;

extern int processedCnt;

aclError SaveBinPostprocess()
{
    aclError retVal;

    LOG("save batch %d start\n", processedCnt);
    DataFrame dataframe = outputDataframe;
    std::vector<std::string>& inferFile_vec = outputDataframe.fileNames;
    aclmdlDataset* output = dataframe.dataset;

    std::string resultFolder = cfg.outDir + "/" + cfg.modelType;
	LOG("resultFolder is %s.\n",resultFolder.c_str());
    DIR* op = opendir(resultFolder.c_str());
    if (NULL == op){
        mkdir(resultFolder.c_str(), 00775);
    }else{
        closedir(op);
    }
	
	

    for (size_t i = 0; i < cfg.outputNum; ++i)
    {        
        aclDataBuffer* dataBuffer = aclmdlGetDatasetBuffer(output, i);
        void* data = aclGetDataBufferAddr(dataBuffer);
		uint32_t length = aclGetDataBufferSize(dataBuffer);
        uint32_t len;
		void* outHostData = NULL;
		uint32_t size = 0;
        len = cfg.outputInfo[i].size;
		LOG("====SaveBinPostprocess.dataset output,outputData:%p,size:%p\n", output, data);        
        LOG("cfg.outputNum is %d, output[%d] real data len %d %d\n",cfg.outputNum, i, len,length);

        ret = aclrtMallocHost(&outHostData, len);
        if (ret != ACL_ERROR_NONE) {
            LOG("Malloc host failed.\n");
            return 1;
        }

        ret = aclrtMemcpy(outHostData, len, data, len, ACL_MEMCPY_DEVICE_TO_HOST);
        if (ret != ACL_ERROR_NONE) {
            LOG("Copy device to host failed.\n");
            aclrtFreeHost(outHostData);
            return 1;
        }

        uint32_t eachSize = len / cfg.batchSize;
        //LOG("inferenceIndex=%d, out_index=%d, len=%d, eachSize=%d\n", inferenceIndex, i, len,eachSize);

			for (size_t j = 0; j < inferFile_vec.size(); j++)
	        {
	            FILE* outputFile;
	            std::string framename = inferFile_vec[j];
	            std::size_t dex = (framename).find_first_of(".");
	            std::string inputFileName = (framename).erase(dex);

	            
	            if (cfg.modelType.compare(0, 6, "resnet") == 0){
	                outputFile = fopen((resultFolder + "/" + "davinci_" + inputFileName + "_"  + "output" + ".bin").c_str(), "wb");
	            }else{
	                outputFile = fopen((resultFolder + "/" + "davinci_" + inputFileName + "_"  + "output" + std::to_string(i) + ".bin").c_str(), "wb");
	            }
	            
	            if (NULL == outputFile){
	                aclrtFreeHost(outHostData);
	                return 1;
	            }

	            fwrite((uint8_t *)outHostData + size, eachSize, sizeof(char), outputFile);
	            fclose(outputFile);
				size = size + eachSize;
	        }

        
        
        ret = aclrtFreeHost(outHostData);
        if (ret != ACL_ERROR_NONE) {
            LOG("Free output host failed.\n");
        }
    }
    
    (void)DestroyDatasetResurce(outputDataframe.dataset, 0);
    
    LOG("save batch %d done\n", processedCnt);
    return ACL_ERROR_NONE;
}

File: ACL_TensorFlow/contrib/recommendation/DIEN_for_ACL/scripts/post_process.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import tensorflow as tf
import numpy as np
import sys
import os

config = tf.ConfigProto()

label_path = sys.argv[1]
npu_predict = sys.argv[2]

final_output = tf.placeholder(shape=[None,2],dtype=tf.float32,name='final_output')
target_ph =tf.placeholder(shape=[None,2],dtype=tf.float32,name='target_ph')

y_hat = tf.add(final_output,0.00000001)
accuracy = tf.reduce_mean(tf.cast(tf.equal(tf.round(y_hat), target_ph), tf.float32))

acc_list = []
init = tf.global_variables_initializer()
with tf.Session(config=config) as sess:
    file_list = os.listdir(label_path)
    file_list.sort()
    for file in file_list:
        if file.endswith('.bin'):
            label = np.fromfile(os.path.join(label_path,file), dtype='float32')
            label = label.reshape(int(label.shape[0]/2),2)
            npu_res = np.fromfile(os.path.join(npu_predict,"davinci_{}_output0.bin".format(file.split('.bin')[0])), dtype='float32')
            npu_res = npu_res.reshape(int(npu_res.shape[0]/2),2)
            acc = sess.run(accuracy,feed_dict={final_output:npu_res,target_ph:label})
            acc_list.append(acc.astype('float32'))
    print("Mean Accuracy: {}".format(sum(acc_list)/len(acc_list)))
